Drag and Drop TabPane can't be used in FXML

Drag and Drop Support for TabPane uses DndTabPaneFactory instead of a subclass. This prevents the use in FXML.

I fixed this issue locally, but i don't know how to build a jar for org.eclipse.fx.ui.controls in order to test my changes.

When i run maven install on the pom.xml in eclipse I get an exception (see below). I have no experience with maven, I am only used to gradle, how can I do this?

```
[ERROR] Cannot resolve project dependencies:
[ERROR]   Software being installed: org.eclipse.fx.ui.controls 3.0.0.qualifier
[ERROR]   Missing requirement: org.eclipse.fx.ui.controls 3.0.0.qualifier requires 'package org.eclipse.fx.core 3.0.0' but it could not be found
[ERROR] 
[ERROR] See http://wiki.eclipse.org/Tycho/Dependency_Resolution_Troubleshooting for help.
[ERROR] Cannot resolve dependencies of MavenProject: org.eclipse.fx.runtime:org.eclipse.fx.ui.controls:3.0.0-SNAPSHOT @ C:\Users\Adrian\Programme\ws-neon\org.eclipse.efxclipse\bundles\runtime\org.eclipse.fx.ui.controls\pom.xml: See log for details -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MavenExecutionException
```

--- a/bundles/runtime/org.eclipse.fx.ui.controls/src/org/eclipse/fx/ui/controls/tabpane/DndTabPane.java
+++ b/bundles/runtime/org.eclipse.fx.ui.controls/src/org/eclipse/fx/ui/controls/tabpane/DndTabPane.java
@@ -19,6 +19,7 @@ import org.eclipse.fx.ui.controls.tabpane.DndTabPaneFactory.DragSetup;
 import org.eclipse.fx.ui.controls.tabpane.DndTabPaneFactory.FeedbackType;
 import org.eclipse.fx.ui.controls.tabpane.skin.DnDTabPaneSkinHookerFullDrag;
 import org.eclipse.fx.ui.controls.tabpane.skin.DndTabPaneSkinHooker;
+import org.eclipse.jdt.annotation.Nullable;
 
 import javafx.beans.property.BooleanProperty;
 import javafx.beans.property.ObjectProperty;
@@ -33,8 +34,8 @@ import javafx.scene.control.TabPane;
  */
 public class DndTabPane extends TabPane {
 	private static final FeedbackType DEFAULT_FEEDBACK_TYPE = FeedbackType.MARKER;
-	private final BooleanProperty allowDetach = new SimpleBooleanProperty(this, "allowDetach"); //$NON-NLS-1$
-	private final ObjectProperty<FeedbackType> feedbackType = new SimpleObjectProperty<>(this, "feedbackType", DEFAULT_FEEDBACK_TYPE); //$NON-NLS-1$
+	private @Nullable BooleanProperty allowDetach;
+	private @Nullable ObjectProperty<FeedbackType> feedbackType;
 
 	private DragSetup setup;
 
@@ -56,7 +57,7 @@ public class DndTabPane extends TabPane {
 	public DndTabPane(FeedbackType feedbackType, boolean allowDetach) {
 		setFeedbackType(feedbackType);
 		setAllowDetach(allowDetach);
-		initListeners(setup -> setup(this.feedbackType::get, this, getChildren(), setup, null));
+		initListeners(setup -> setup(feedbackTypeProperty()::get, this, getChildren(), setup, null));
 	}
 
 	/**
@@ -96,7 +97,7 @@ public class DndTabPane extends TabPane {
 				setupDnd(newValue, setup);
 			}
 		});
-		this.allowDetach.addListener((observable, oldValue, newValue) -> {
+		allowDetachProperty().addListener((observable, oldValue, newValue) -> {
 			if (newValue != null) {
 				Skin<?> skin = getSkin();
 				if (skin != null) {
@@ -122,6 +123,9 @@ public class DndTabPane extends TabPane {
 	 * @return a property indicating whether Tabs can be detached or not
 	 */
 	public BooleanProperty allowDetachProperty() {
+		if (this.allowDetach == null) {
+			this.allowDetach = new SimpleBooleanProperty(this, "allowDetach"); //$NON-NLS-1$
+		}
 		return this.allowDetach;
 	}
 
@@ -129,7 +133,7 @@ public class DndTabPane extends TabPane {
 	 * @return whether Tabs can be detached or not
 	 */
 	public boolean isAllowDetach() {
-		return this.allowDetach.get();
+		return this.allowDetach != null ? this.allowDetach.get() : false;
 	}
 
 	/**
@@ -137,13 +141,16 @@ public class DndTabPane extends TabPane {
 	 *            whether Tabs can be detached or not
 	 */
 	public void setAllowDetach(boolean allowDetach) {
-		this.allowDetach.set(allowDetach);
+		allowDetachProperty().set(allowDetach);
 	}
 
 	/**
 	 * @return the type of visual Feedback for the User
 	 */
 	public ObjectProperty<FeedbackType> feedbackTypeProperty() {
+		if (this.feedbackType == null) {
+			this.feedbackType = new SimpleObjectProperty<>(this, "feedbackType", DEFAULT_FEEDBACK_TYPE); //$NON-NLS-1$
+		}
 		return this.feedbackType;
 	}
 
@@ -151,7 +158,7 @@ public class DndTabPane extends TabPane {
 	 * @return the type of visual Feedback for the User
 	 */
 	public FeedbackType getFeedbackType() {
-		return this.feedbackType.get();
+		return this.feedbackType != null ? this.feedbackType.get() : DEFAULT_FEEDBACK_TYPE;
 	}
 
 	/**
@@ -159,6 +166,6 @@ public class DndTabPane extends TabPane {
 	 *            the type of visual Feedback for the User
 	 */
 	public void setFeedbackType(FeedbackType feedbackType) {
-		this.feedbackType.set(feedbackType);
+		feedbackTypeProperty().set(feedbackType);
 	}
 }

--- a/bundles/runtime/org.eclipse.fx.ui.controls/src/org/eclipse/fx/ui/controls/tabpane/DndTabPane.java
+++ b/bundles/runtime/org.eclipse.fx.ui.controls/src/org/eclipse/fx/ui/controls/tabpane/DndTabPane.java
@@ -92,6 +92,9 @@ public class DndTabPane extends TabPane {
 			if (newValue != null) {
 				Skin<?> skin = getSkin();
 				if (skin != null) {
+					if (this.setup != null) {
+						teardown(this.setup);
+					}
 					setupDnd(skin, setup);
 				}
 			}

--- a/bundles/runtime/org.eclipse.fx.ui.controls/src/org/eclipse/fx/ui/controls/tabpane/DndTabPane.java
+++ b/bundles/runtime/org.eclipse.fx.ui.controls/src/org/eclipse/fx/ui/controls/tabpane/DndTabPane.java
@@ -33,8 +33,8 @@ import javafx.scene.control.TabPane;
  */
 public class DndTabPane extends TabPane {
 	private static final FeedbackType DEFAULT_FEEDBACK_TYPE = FeedbackType.MARKER;
-	private final BooleanProperty allowDetach = new SimpleBooleanProperty();
-	private final ObjectProperty<FeedbackType> feedbackType = new SimpleObjectProperty<>(DEFAULT_FEEDBACK_TYPE);
+	private final BooleanProperty allowDetach = new SimpleBooleanProperty(this, "allowDetach"); //$NON-NLS-1$
+	private final ObjectProperty<FeedbackType> feedbackType = new SimpleObjectProperty<>(this, "feedbackType", DEFAULT_FEEDBACK_TYPE); //$NON-NLS-1$
 
 	private DragSetup setup;
 

--- a/bundles/runtime/org.eclipse.fx.ui.controls/src/org/eclipse/fx/ui/controls/tabpane/DndTabPane.java
+++ b/bundles/runtime/org.eclipse.fx.ui.controls/src/org/eclipse/fx/ui/controls/tabpane/DndTabPane.java
@@ -31,6 +31,8 @@ import javafx.scene.control.TabPane;
 
 /**
  * A TabPane with Drag and Drop Support
+ *
+ * @since 3.0.0
  */
 public class DndTabPane extends TabPane {
 	private static final FeedbackType DEFAULT_FEEDBACK_TYPE = FeedbackType.MARKER;

--- a/bundles/runtime/org.eclipse.fx.ui.controls/src/org/eclipse/fx/ui/controls/tabpane/DndTabPane.java
+++ b/bundles/runtime/org.eclipse.fx.ui.controls/src/org/eclipse/fx/ui/controls/tabpane/DndTabPane.java
@@ -1,0 +1,153 @@
+package org.eclipse.fx.ui.controls.tabpane;
+
+import static org.eclipse.fx.ui.controls.tabpane.DndTabPaneFactory.setup;
+import static org.eclipse.fx.ui.controls.tabpane.DndTabPaneFactory.teardown;
+
+import java.util.function.Consumer;
+
+import org.eclipse.fx.ui.controls.tabpane.DndTabPaneFactory.DragSetup;
+import org.eclipse.fx.ui.controls.tabpane.DndTabPaneFactory.FeedbackType;
+import org.eclipse.fx.ui.controls.tabpane.skin.DnDTabPaneSkinHookerFullDrag;
+import org.eclipse.fx.ui.controls.tabpane.skin.DndTabPaneSkinHooker;
+
+import javafx.beans.property.BooleanProperty;
+import javafx.beans.property.ObjectProperty;
+import javafx.beans.property.SimpleBooleanProperty;
+import javafx.beans.property.SimpleObjectProperty;
+import javafx.scene.control.Skin;
+import javafx.scene.control.SkinBase;
+import javafx.scene.control.TabPane;
+
+/**
+ * A TabPane with Drag and Drop Support
+ *
+ * @author Adrodoc55
+ */
+public class DndTabPane extends TabPane {
+	private static final FeedbackType DEFAULT_FEEDBACK_TYPE = FeedbackType.MARKER;
+	private final BooleanProperty allowDetach = new SimpleBooleanProperty();
+	private final ObjectProperty<FeedbackType> feedbackType = new SimpleObjectProperty<>(DEFAULT_FEEDBACK_TYPE);
+
+	private DragSetup setup;
+
+	/**
+	 * Create a tab pane with a default setup for Drag and Drop Feedback
+	 */
+	public DndTabPane() {
+		this(DEFAULT_FEEDBACK_TYPE, false);
+	}
+
+	/**
+	 * Create a tab pane with a default setup for Drag and Drop Feedback
+	 *
+	 * @param feedbackType
+	 *            the feedback type
+	 * @param allowDetach
+	 *            allow detaching
+	 */
+	public DndTabPane(FeedbackType feedbackType, boolean allowDetach) {
+		setFeedbackType(feedbackType);
+		setAllowDetach(allowDetach);
+		initListeners(setup -> setup(this.feedbackType::get, this, getChildren(), setup, null));
+	}
+
+	/**
+	 * Create a tab pane and set the drag strategy
+	 *
+	 * @param allowDetach
+	 *            allow detaching
+	 * @param setup
+	 *            the setup instance for the pane
+	 */
+	public DndTabPane(boolean allowDetach, Consumer<DragSetup> setup) {
+		this(DEFAULT_FEEDBACK_TYPE, allowDetach, setup);
+	}
+
+	/**
+	 * Create a tab pane and set the drag strategy
+	 *
+	 * @param feedbackType
+	 *            the feedback type
+	 * @param allowDetach
+	 *            allow detaching
+	 * @param setup
+	 *            the setup instance for the pane
+	 */
+	public DndTabPane(FeedbackType feedbackType, boolean allowDetach, Consumer<DragSetup> setup) {
+		setFeedbackType(feedbackType);
+		setAllowDetach(allowDetach);
+		initListeners(setup);
+	}
+
+	private void initListeners(Consumer<DragSetup> setup) {
+		skinProperty().addListener((observable, oldValue, newValue) -> {
+			if (oldValue != null && this.setup != null) {
+				teardown(this.setup);
+			}
+			if (newValue != null) {
+				setupDnd(newValue, setup);
+			}
+		});
+		this.allowDetach.addListener((observable, oldValue, newValue) -> {
+			if (newValue != null) {
+				Skin<?> skin = getSkin();
+				if (skin != null) {
+					setupDnd(skin, setup);
+				}
+			}
+		});
+	}
+
+	private void setupDnd(Skin<?> skin, Consumer<DragSetup> setup) {
+		if (isAllowDetach()) {
+			this.setup = new DnDTabPaneSkinHookerFullDrag((SkinBase<TabPane>) skin);
+		} else {
+			this.setup = new DndTabPaneSkinHooker((Skin<TabPane>) skin);
+		}
+		setup.accept(this.setup);
+	}
+
+	/**
+	 * @return a property indicating whether Tabs can be detached or not
+	 */
+	public BooleanProperty allowDetachProperty() {
+		return this.allowDetach;
+	}
+
+	/**
+	 * @return whether Tabs can be detached or not
+	 */
+	public boolean isAllowDetach() {
+		return this.allowDetach.get();
+	}
+
+	/**
+	 * @param allowDetach
+	 *            whether Tabs can be detached or not
+	 */
+	public void setAllowDetach(boolean allowDetach) {
+		this.allowDetach.set(allowDetach);
+	}
+
+	/**
+	 * @return the type of visual Feedback for the User
+	 */
+	public ObjectProperty<FeedbackType> feedbackTypeProperty() {
+		return this.feedbackType;
+	}
+
+	/**
+	 * @return the type of visual Feedback for the User
+	 */
+	public FeedbackType getFeedbackType() {
+		return this.feedbackType.get();
+	}
+
+	/**
+	 * @param feedbackType
+	 *            the type of visual Feedback for the User
+	 */
+	public void setFeedbackType(FeedbackType feedbackType) {
+		this.feedbackType.set(feedbackType);
+	}
+}

--- a/bundles/runtime/org.eclipse.fx.ui.controls/src/org/eclipse/fx/ui/controls/tabpane/DndTabPane.java
+++ b/bundles/runtime/org.eclipse.fx.ui.controls/src/org/eclipse/fx/ui/controls/tabpane/DndTabPane.java
@@ -1,3 +1,13 @@
+/*******************************************************************************
+ * Copyright (c) 2014 BestSolution.at and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Adrodoc55<adrodoc55@googlemail.com> - initial API and implementation
+ *******************************************************************************/
 package org.eclipse.fx.ui.controls.tabpane;
 
 import static org.eclipse.fx.ui.controls.tabpane.DndTabPaneFactory.setup;
@@ -20,8 +30,6 @@ import javafx.scene.control.TabPane;
 
 /**
  * A TabPane with Drag and Drop Support
- *
- * @author Adrodoc55
  */
 public class DndTabPane extends TabPane {
 	private static final FeedbackType DEFAULT_FEEDBACK_TYPE = FeedbackType.MARKER;

--- a/bundles/runtime/org.eclipse.fx.ui.controls/src/org/eclipse/fx/ui/controls/tabpane/DndTabPaneFactory.java
+++ b/bundles/runtime/org.eclipse.fx.ui.controls/src/org/eclipse/fx/ui/controls/tabpane/DndTabPaneFactory.java
@@ -12,22 +12,21 @@ package org.eclipse.fx.ui.controls.tabpane;
 
 import java.util.function.Consumer;
 import java.util.function.Function;
+import java.util.function.Supplier;
 
 import org.eclipse.fx.core.SystemUtils;
 import org.eclipse.fx.ui.controls.dnd.EFXDragEvent;
 import org.eclipse.fx.ui.controls.markers.PositionMarker;
 import org.eclipse.fx.ui.controls.markers.TabOutlineMarker;
-import org.eclipse.fx.ui.controls.tabpane.skin.DnDTabPaneSkinHookerFullDrag;
 import org.eclipse.fx.ui.controls.tabpane.skin.DndTabPaneSkinHooker;
 import org.eclipse.jdt.annotation.NonNull;
 import org.eclipse.jdt.annotation.Nullable;
 
+import javafx.collections.ObservableList;
 import javafx.event.Event;
 import javafx.geometry.BoundingBox;
 import javafx.geometry.Bounds;
 import javafx.scene.Node;
-import javafx.scene.control.Skin;
-import javafx.scene.control.SkinBase;
 import javafx.scene.control.TabPane;
 import javafx.scene.input.DragEvent;
 import javafx.scene.layout.Pane;
@@ -56,7 +55,7 @@ public final class DndTabPaneFactory {
 	 */
 	public static <D extends Node & DragSetup> Pane setup(FeedbackType feedbackType, D dragSetup, Consumer<GenericTab> detachHandler) {
 		StackPane pane = new StackPane();
-		setup(feedbackType, pane, dragSetup, detachHandler);
+		setup(() -> feedbackType, pane, pane.getChildren(), dragSetup, detachHandler);
 		pane.getChildren().add(dragSetup);
 		return pane;
 	}
@@ -77,12 +76,10 @@ public final class DndTabPaneFactory {
 	 *
 	 * @param feedbackType
 	 *            the feedback type
-	 * @param setup
-	 *            consumer to set up the tab pane
 	 * @return a pane containing the TabPane
 	 */
-	public static Pane createDefaultDnDPane(FeedbackType feedbackType, Consumer<TabPane> setup) {
-		return createDefaultDnDPane(feedbackType, false, setup);
+	public static TabPane createDefaultDnDPane(FeedbackType feedbackType) {
+		return createDefaultDnDPane(feedbackType, false);
 	}
 
 	/**
@@ -91,24 +88,11 @@ public final class DndTabPaneFactory {
 	 * @param setup
 	 *            the setup instance for the pane
 	 * @param allowDetach
-	 * 			  allow detaching
+	 *            allow detaching
 	 * @return the tab pane
 	 */
 	public static TabPane createDndTabPane(Consumer<DragSetup> setup, boolean allowDetach) {
-		return new TabPane() {
-			@Override
-			protected javafx.scene.control.Skin<?> createDefaultSkin() {
-				Skin<?> skin = super.createDefaultSkin();
-				DragSetup ds;
-				if( allowDetach ) {
-					ds = new DnDTabPaneSkinHookerFullDrag((SkinBase<TabPane>)skin);
-				} else {
-					ds = new DndTabPaneSkinHooker((Skin<TabPane>) skin);
-				}
-				setup.accept(ds);
-				return skin;
-			}
-		};
+		return new DndTabPane(allowDetach, setup);
 	}
 
 	/**
@@ -118,35 +102,14 @@ public final class DndTabPaneFactory {
 	 *            the feedback type
 	 * @param allowDetach
 	 *            allow detaching
-	 * @param setup
-	 *            consumer to set up the tab pane
-	 * @return a pane containing the TabPane
+	 * @return the TabPane
 	 */
-	public static Pane createDefaultDnDPane(FeedbackType feedbackType, boolean allowDetach, Consumer<TabPane> setup) {
-		StackPane pane = new StackPane();
-		TabPane tabPane;
-		if( SystemUtils.isFX9() || Boolean.getBoolean("java9.tabpane") ) {
-			tabPane = new TabPane();
+	public static TabPane createDefaultDnDPane(FeedbackType feedbackType, boolean allowDetach) {
+		if (SystemUtils.isFX9() || Boolean.getBoolean("java9.tabpane")) {
+			return new TabPane();
 		} else {
-			tabPane = new TabPane() {
-				@Override
-				protected javafx.scene.control.Skin<?> createDefaultSkin() {
-					Skin<?> skin = super.createDefaultSkin();
-					DragSetup ds;
-					if( allowDetach ) {
-						ds = new DnDTabPaneSkinHookerFullDrag((SkinBase<TabPane>)skin);
-					} else {
-						ds = new DndTabPaneSkinHooker((Skin<TabPane>) skin);
-					}
-					setup(feedbackType, pane, ds, null);
-					return skin;
-				}
-			};
+			return new DndTabPane(feedbackType, allowDetach);
 		}
-
-		setup.accept(tabPane);
-		pane.getChildren().add(tabPane);
-		return pane;
 	}
 
 	/**
@@ -186,15 +149,30 @@ public final class DndTabPaneFactory {
 	 *
 	 * @param layoutNode
 	 *            the layout node used to position
+	 * @param layoutNodeChildren
+	 *            the children of {@code layoutNode}
 	 * @param setup
 	 *            the setup
 	 */
 	@SuppressWarnings("null")
-	static void setup(FeedbackType type, Pane layoutNode, DragSetup setup, Consumer<GenericTab> detachHandler) {
+	static void setup(Supplier<FeedbackType> typeProvider, Node layoutNode, ObservableList<Node> layoutNodeChildren, DragSetup setup, Consumer<GenericTab> detachHandler) {
 		setup.setStartFunction((t) -> Boolean.TRUE);
-		setup.setFeedbackConsumer((d) -> handleFeedback(type, layoutNode, d));
+		setup.setFeedbackConsumer((d) -> handleFeedback(typeProvider, layoutNode, layoutNodeChildren, d));
 		setup.setDropConsumer(d -> handleDropped(d, detachHandler));
 		setup.setDragFinishedConsumer(DndTabPaneFactory::handleFinished);
+	}
+
+	/**
+	 * Teardown insert marker
+	 *
+	 * @param setup
+	 *            the setup
+	 */
+	static void teardown(DragSetup setup) {
+		setup.setStartFunction(null);
+		setup.setFeedbackConsumer(null);
+		setup.setDropConsumer(null);
+		setup.setDragFinishedConsumer(null);
 	}
 
 	private static void handleDropped(DroppedData data, Consumer<GenericTab> detachHandler) {
@@ -219,7 +197,7 @@ public final class DndTabPaneFactory {
 		}
 	}
 
-	private static void handleFeedback(FeedbackType type, Pane layoutNode, FeedbackData data) {
+	private static void handleFeedback(Supplier<FeedbackType> typeProvider, Node layoutNode, ObservableList<Node> layoutNodeChildren, FeedbackData data) {
 		if (data.dropType == DropType.NONE) {
 			cleanup();
 			return;
@@ -228,10 +206,10 @@ public final class DndTabPaneFactory {
 		MarkerFeedback f = CURRENT_FEEDBACK;
 		if (f == null || !f.data.equals(data)) {
 			cleanup();
-			if (type == FeedbackType.MARKER) {
-				CURRENT_FEEDBACK = handleMarker(layoutNode, data);
+			if (typeProvider.get() == FeedbackType.OUTLINE) {
+				CURRENT_FEEDBACK = handleOutline(layoutNode, layoutNodeChildren, data);
 			} else {
-				CURRENT_FEEDBACK = handleOutline(layoutNode, data);
+				CURRENT_FEEDBACK = handleMarker(layoutNodeChildren, data);
 			}
 		}
 	}
@@ -247,9 +225,9 @@ public final class DndTabPaneFactory {
 		}
 	}
 
-	private static MarkerFeedback handleMarker(Pane layoutNode, FeedbackData data) {
+	private static MarkerFeedback handleMarker(ObservableList<Node> layoutNodeChildren, FeedbackData data) {
 		PositionMarker marker = null;
-		for (Node n : layoutNode.getChildren()) {
+		for (Node n : layoutNodeChildren) {
 			if (n instanceof PositionMarker) {
 				marker = (PositionMarker) n;
 			}
@@ -258,7 +236,7 @@ public final class DndTabPaneFactory {
 		if (marker == null) {
 			marker = new PositionMarker();
 			marker.setManaged(false);
-			layoutNode.getChildren().add(marker);
+			layoutNodeChildren.add(marker);
 		} else {
 			marker.setVisible(true);
 		}
@@ -292,10 +270,10 @@ public final class DndTabPaneFactory {
 	}
 
 	@SuppressWarnings("null")
-	private static MarkerFeedback handleOutline(Pane layoutNode, FeedbackData data) {
+	private static MarkerFeedback handleOutline(Node layoutNode, ObservableList<Node> layoutNodeChildren, FeedbackData data) {
 		TabOutlineMarker marker = null;
 
-		for (Node n : layoutNode.getChildren()) {
+		for (Node n : layoutNodeChildren) {
 			if (n instanceof TabOutlineMarker) {
 				marker = (TabOutlineMarker) n;
 			}
@@ -305,7 +283,7 @@ public final class DndTabPaneFactory {
 			marker = new TabOutlineMarker(layoutNode.getBoundsInLocal(), new BoundingBox(data.bounds.getMinX(), data.bounds.getMinY(), data.bounds.getWidth(), data.bounds.getHeight()), data.dropType == DropType.BEFORE);
 			marker.setManaged(false);
 			marker.setMouseTransparent(true);
-			layoutNode.getChildren().add(marker);
+			layoutNodeChildren.add(marker);
 		} else {
 			marker.updateBounds(layoutNode.getBoundsInLocal(), new BoundingBox(data.bounds.getMinX(), data.bounds.getMinY(), data.bounds.getWidth(), data.bounds.getHeight()), data.dropType == DropType.BEFORE);
 			marker.setVisible(true);

